### ISSUE TITLE
white lines in do blocks

### ISSUE THREAD

**Issue by xubin990510** [itchyny/vim-haskell-indent#8]
like this: 

```
main = do
    string <- getLine

    print string
```

doing `gg=G` will make it indented like this:

```
main = do
    string <- getLine

print string
```

**Comment by itchyny** [comment]
This is actually a difficult problem. When I modify the code to fix this, we would have trouble in defining a function after a do block.

**Comment by xubin990510** [comment]
Is it possible to detect function  definition? (Quite new to Haskell. If that doesn't make sense don't mind ^^)

**Comment by itchyny** [comment]
Quite a difficult problem. Assuming that the indentation of the monad continues,
```haskell
main = do
  x <- getLine

  f g
    h
    i
    = g h i
```
oops, now it is a function definition. We have to check the succeeding lines which may contain complex syntaxes like nested let..in-s, argument matching and guards.

For another example example, see the following indent-broken code,
```haskell
main = do
let x = 10
f = ($)
y = f
print 1
```
the corrected code is as follows.
```haskell
main = do
  let x = 10
      f = ($)
      y = f
  print 1
```
but if there is the `in` clause,
```haskell
main = do
let x = 10
in print x
f = ($)
y = f
print 1
```
the corrected code is as follows.
```haskell
main = do
  let x = 10
      in print x
f = ($)
y = f
  print 1
```
The plugin cannot deal with both cases, yeah, I should fix for some patterns...

Basically it is impossible to create a comfortable indentation plugin which allows us to write Haskell codes with no indentation space keys and no backspace keys removing the indentation. Talking of auto-indentation (like `gg=G`), it might not impossible, but difficult in my opinion.

**Comment by xubin990510** [comment]
OK, I got the point. Thanks!
